contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,952 @@
+     * @param _amount            Units of asset to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override onlyVault nonReentrant {
+        require(_asset == WETH, "Unsupported asset");
+        require(_recipient == vaultAddress, "Only withdraw to vault allowed");
+
+        uint256 wethBalance = IERC20(WETH).balanceOf(address(this));
+        if (wethBalance < _amount) {
+            uint256 additionalWethRequired = _amount - wethBalance;
+            (uint256 wethInThePool, ) = getPositionPrincipal();
+
+            if (wethInThePool < additionalWethRequired) {
+                revert NotEnoughWethLiquidity(
+                    wethInThePool,
+                    additionalWethRequired
+                );
+            }
+
+            uint256 shareOfWethToRemove = additionalWethRequired.divPrecisely(
+                wethInThePool
+            ) + 1;
+            _removePartialLiquidity(shareOfWethToRemove);
+        }
+
+        // burn remaining OETHb
+        _burnOethbOnTheContract();
+        _withdraw(_recipient, _amount);
+    }
+
+    /**
+     * @notice Withdraw all supported assets from platform and
+     * sends to the OToken's Vault.
+     */
+    function withdrawAll() external override onlyVault nonReentrant {
+        _removeAllLiquidity();
+
+        uint256 balance = IERC20(WETH).balanceOf(address(this));
+        if (balance > 0) {
+            _withdraw(vaultAddress, balance);
+        }
+        // burn remaining OETHb
+        _burnOethbOnTheContract();
+    }
+
+    function _withdraw(address _recipient, uint256 _amount) internal {
+        require(_amount > 0, "Must withdraw something");
+        require(_recipient != address(0), "Must specify recipient");
+
+        IERC20(WETH).safeTransfer(_recipient, _amount);
+        emit Withdrawal(WETH, address(0), _amount);
+    }
+
+    /**
+     * @dev Collect the AERO token from the gauge
+     */
+    function _collectRewardTokens() internal override {
+        clGauge.getReward(tokenId);
+        super._collectRewardTokens();
+    }
+
+    /**
+     * @dev Retuns bool indicating whether asset is supported by strategy
+     * @param _asset Address of the asset
+     */
+    function supportsAsset(address _asset) public view override returns (bool) {
+        return _asset == WETH;
+    }
+
+    /**
+     * @dev Internal method to respond to the addition of new asset / pTokens
+            We need to give the Aerodrome pool approval to transfer the
+            asset.
+     * @param _asset Address of the asset to approve
+     * @param _pToken Address of the pToken
+     */
+    // solhint-disable-next-line no-unused-vars
+    function _abstractSetPToken(address _asset, address _pToken)
+        internal
+        override
+    {
+        IERC20(_asset).safeApprove(address(positionManager), type(uint256).max);
+        IERC20(_asset).safeApprove(address(swapRouter), type(uint256).max);
```

Comment:
`setPTokenAddress` and `removePToken` methods always revert but this doesn't seem to do that. 

`_setPTokenAddress` is only called in initialize if any `assets` exists. As far as this strategy is concerned, it's always gonna be an empty array (as hardcoded in the initialize). I don't see any reason to do these approvals here (or this method to not revert like its external counterpart)

Reply:
good point: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/93d5b964117046cafb237370997b99ae0aacf13e